aten/src/TH/vector/AVX2.c

```diff
@@ -45,4 +47,62 @@ void THFloatVector_cadd_AVX2(float *z, const float *x, const float *y, const flo
   }
 }
 
+static void normal_fill_16_AVX2(float *data,
+                                const __m256* two_pi,
+                                const __m256* one,
+                                const __m256* minus_two,
+                                const __m256* mean,
+                                const __m256* stddev) {
+  const __m256 u1 = _mm256_sub_ps(*one, _mm256_loadu_ps(data));
+  const __m256 u2 = _mm256_loadu_ps(data + 8);
+
+  // sincos256_ps and log256_ps are from avx_mathfun.h
+  const __m256 radius = _mm256_sqrt_ps(_mm256_mul_ps(*minus_two, log256_ps(u1)));
+  const __m256 theta = _mm256_mul_ps(*two_pi, u2);
+
+  __m256 sintheta, costheta;
+  sincos256_ps(theta, &sintheta, &costheta);
+
+  const __m256 n1 = _mm256_mul_ps(radius, costheta);
+  const __m256 n2 = _mm256_mul_ps(radius, sintheta);
+
+  _mm256_storeu_ps(data, _mm256_fmadd_ps(n1, *stddev, *mean));
+  _mm256_storeu_ps(data + 8, _mm256_fmadd_ps(n2, *stddev, *mean));
+}
+
+void THFloatVector_normal_fill_AVX2(float *data,
+                                    const int64_t size,
+                                    THGenerator *generator,
+                                    const float mean,
+                                    const float stddev)
+{
+  THAssert(size >= 16 && "Size must be >= 16 for AVX2 normal fill");
+  const __m256 two_pi = _mm256_set1_ps(2.0f * M_PI);
+  const __m256 one = _mm256_set1_ps(1.0f);
+  const __m256 minus_two = _mm256_set1_ps(-2.0f);
+  const __m256 mean_v = _mm256_set1_ps(mean);
+  const __m256 stddev_v = _mm256_set1_ps(stddev);
+
+  // First fill the data with the uniform numbers. Box-Mueller is a 2 -> 2
+  // mapping of 2 uniform numbers to 2 normal numbers (per iteration), so we
+  // we need exactly as much space for uniform and normal numbers and can just
+  // use the single buffer for both.
+  for (int64_t i = 0; i < size; ++i) {
+    data[i] = THRandom_uniformFloat(generator, 0, 1);
+  }
+
+  for (int64_t i = 0; i < size - 15; i += 16) {
+    normal_fill_16_AVX2(data + i, &two_pi, &one, &minus_two, &mean_v, &stddev_v);
+  }
+
+  if (size % 16 != 0) {
+    // We rewind so that we have 16 values and then compute them in one step.
+    data += (size - 16);
```

Comment:
I think in here `data` still points to the beginning of the buffer (you've been doing `data + i`, not `data += 16` in the loop above). Shouldn't this be `data = data + size - 16`? It might be a tiny bit cleaner and safer to declare that as a new pointer instead of overwriting this one

Reply:
You're right that it's still pointing to the start. The idea here is to place it to the start of the last 16 values. So I think it's ok, but I'll apply your suggestion of `data = data + size - 16`.

Reply:
Ohhh right. Not sure why I thought it was that way